### PR TITLE
Implement holiday schedules

### LINKED ISSUE #17392: [SQA] Door Lock - Missing commands has to be implemented.
**Issue** : Door lock cluster has to be implemenated with below commands based on test plan.

**commit used** : fa5561c23d8c1f9464da8669931affbae0660f51		

**Referance** :  https://github.com/CHIP-Specifications/chip-test-plans/blob/master/src/cluster/door_lock_Cluster.adoc#2-test-case-list

**List of missing command based on the test cases** :

- [TC-DL-2.4] - Unlock with Timeout command
- [TC-DL-2.5] - Get Log Record, Get Log Record Response Command
- [TC-DL-2.6] - Set PIN Code, Get PIN Code, Get PIN Code Response, Clear PIN Code, Clear All PIN Codes
- [TC-DL-2.7] - Set User Status, Get User Status , Get User Status Response
- [TC-DL-2.9] - Set Holiday Schedule, Get Holiday Schedule, Get Holiday Schedule Response , Clear Holiday Schedule
- [TC-DL-2.10]- Set User Type, Get User Type ,Get User Type Response
- [TC-DL-2.12]- Set RFID Code, Get RFID Code, Get RFID Code Response, Clear RFID Code and Clear All RFID Codes
- [TC-DL-2.15]- Operation Event Notification Command
- [TC-DL-2.16]- Programming Event Notification Command

### PR DESCRIPTION
**Pull request by Morozov-5F**
#### Problem
* Fixes #17392 by implementing holiday schedules.

#### Change overview
Implemented new commands for the door lock cluster related to holiday schedules:
* Set Holiday Schedule
* Get Holiday Schedule
* Get Holiday Schedule Response
* Clear Holiday Schedule
Implemented new attribute related to holiday schedules: 
* NumberOfHolidaySchedulesSupported
Updated automated tests for the door lock schedules to include testing of the holiday schedules.

#### Testing
* Ran existing YAML tests (DL_UsersAndCredentials, DL_LockUnlock and DL_Schedules)
* Updated DL_Schedules test to include tests for holiday schedules commands.

### CODE REVIEW

File: examples/lock-app/linux/include/LockEndpoint.h [hunk 1]
```diff
@@ -36,19 +36,22 @@ struct LockUserInfo
 struct LockCredentialInfo;
 struct WeekDaysScheduleInfo;
 struct YearDayScheduleInfo;
+struct HolidayScheduleInfo;
```
Comment (andy31415) [inline examples/lock-app/linux/include/LockEndpoint.h:39]: Why are these forward-declared where as LockUserInfo is declared in place?

**Review by andy31415 — APPROVED** [review]

**Review by tehampson — APPROVED** [review]

File: examples/lock-app/linux/include/LockEndpoint.h (hunk 1 above)
Comment (Morozov-5F) [reply examples/lock-app/linux/include/LockEndpoint.h:39]: Uh, no particular reason except I didn't want to clutter the top of the file with complete declarations.

**Review by woody-apple — APPROVED** [review]

File: examples/lock-app/linux/src/LockManager.cpp [hunk 2]
```diff
@@ -89,15 +89,25 @@ bool LockManager::InitEndpoint(chip::EndpointId endpointId)
...
     }
 
+    uint8_t numberOfHolidaySchedules = 0;
+    if (!DoorLockServer::Instance().GetNumberOfHolidaySchedulesSupported(endpointId, numberOfHolidaySchedules))
+    {
+        ChipLogError(
+            Zcl,
+            "Unable to get number of supported holiday schedules when initializing lock endpoint, defaulting to 10 [endpointId=%d]",
+            endpointId);
+        numberOfYearDaySchedulesPerUser = 10;
```
Comment (bzbarsky-apple) [inline examples/lock-app/linux/src/LockManager.cpp:99]: Why are we resetting `numberOfYearDaySchedulesPerUser` if holiday schedules are not supported?

File: src/app/clusters/door-lock-server/door-lock-server.cpp [hunk 3]
```diff
@@ -2312,6 +2318,72 @@ void DoorLockServer::sendGetYearDayScheduleResponse(chip::app::CommandHandler *
...
 
+bool DoorLockServer::holidayIndexValid(chip::EndpointId endpointId, uint8_t holidayIndex)
+{
+    uint8_t holidaysSupported;
+    if (!GetNumberOfHolidaySchedulesSupported(endpointId, holidaysSupported))
+    {
+        return false;
+    }
+
+    // appclusters, 5.2.4.22-25: year day index changes from 1 to maxNumberOfHolidaySchedules
```
Comment (bzbarsky-apple) [inline src/app/clusters/door-lock-server/door-lock-server.cpp:2329]: What does this have to do with "year day index"?

File: examples/lock-app/linux/src/LockManager.cpp (hunk 2 above)
Comment (Morozov-5F) [reply examples/lock-app/linux/src/LockManager.cpp:99]: This was a typo, fixed here: https://github.com/project-chip/connectedhomeip/pull/18860

File: src/app/clusters/door-lock-server/door-lock-server.cpp (hunk 3 above)
Comment (Morozov-5F) [reply src/app/clusters/door-lock-server/door-lock-server.cpp:2329]: Just a typo, I copy pasted the comment, here's the result. I should probably revise all the comments.
